In [9]:
# %matplotlib widget

from __future__ import annotations

import re
from collections import defaultdict
from dataclasses import dataclass, field
from itertools import permutations, product
from math import inf
from random import choice

import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
import networkx as nx
import numpy as np
import numpy.typing as npt
from mpl_toolkits.mplot3d import axes3d
from numpy import int_, object_
from numpy.typing import NDArray
from test_utilities import run_tests_params
from util import print_hex

COLORS = list(mcolors.CSS4_COLORS.keys())

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc read-aloud"><h2>--- Day 5: Sunny with a Chance of Asteroids ---</h2><p>You're starting to sweat as the ship makes its way toward Mercury.  The Elves suggest that you get the air conditioner working by upgrading your ship computer to support the Thermal Environment Supervision Terminal.</p>
<p>The Thermal Environment Supervision Terminal (TEST) starts by running a <em>diagnostic program</em> (your puzzle input).  The TEST diagnostic program will run on <a href="2">your existing Intcode computer</a> after a few modifications:</p>
<p><em>First</em>, you'll need to add <em>two new instructions</em>:</p>
<ul>
<li>Opcode <code>3</code> takes a single integer as <em>input</em> and saves it to the position given by its only parameter. For example, the instruction <code>3,50</code> would take an input value and store it at address <code>50</code>.</li>
<li>Opcode <code>4</code> <em>outputs</em> the value of its only parameter. For example, the instruction <code>4,50</code> would output the value at address <code>50</code>.</li>
</ul>
<p>Programs that use these instructions will come with documentation that explains what should be connected to the input and output. The program <code>3,0,4,0,99</code> outputs whatever it gets as input, then halts.</p>
<p><em>Second</em>, you'll need to add support for <em>parameter modes</em>:</p>
<p>Each parameter of an instruction is handled based on its parameter mode.  Right now, your ship computer already understands parameter mode <code>0</code>, <em>position mode</em>, which causes the parameter to be interpreted as a <em>position</em> - if the parameter is <code>50</code>, its value is <em>the value stored at address <code>50</code> in memory</em>. Until now, all parameters have been in position mode.</p>
<p>Now, your ship computer will also need to handle parameters in mode <code>1</code>, <em>immediate mode</em>. In immediate mode, a parameter is interpreted as a <em>value</em> - if the parameter is <code>50</code>, its value is simply <em><code>50</code></em>.</p>
<p>Parameter modes are stored in the same value as the instruction's opcode.  The opcode is a two-digit number based only on the ones and tens digit of the value, that is, the opcode is the rightmost two digits of the first value in an instruction. Parameter modes are single digits, one per parameter, read right-to-left from the opcode: the first parameter's mode is in the hundreds digit, the second parameter's mode is in the thousands digit, the third parameter's mode is in the ten-thousands digit, and so on. Any missing modes are <code>0</code>.</p>
<p>For example, consider the program <code>1002,4,3,4,33</code>.</p>
<p>The first instruction, <code>1002,4,3,4</code>, is a <em>multiply</em> instruction - the rightmost two digits of the first value, <code>02</code>, indicate opcode <code>2</code>, multiplication.  Then, going right to left, the parameter modes are <code>0</code> (hundreds digit), <code>1</code> (thousands digit), and <code>0</code> (ten-thousands digit, not present and therefore zero):</p>
<pre><code>ABCDE
 1002

DE - two-digit opcode, 02 == opcode 2
C - mode of 1st parameter, 0 == position mode
B - mode of 2nd parameter, 1 == immediate mode
A - mode of 3rd parameter, 0 == position mode,
omitted due to being a leading zero
</code></pre>

<p>This instruction multiplies its first two parameters.  The first parameter, <code>4</code> in position mode, works like it did before - its value is the value stored at address <code>4</code> (<code>33</code>). The second parameter, <code>3</code> in immediate mode, simply has value <code>3</code>. The result of this operation, <code>33 * 3 = 99</code>, is written according to the third parameter, <code>4</code> in position mode, which also works like it did before - <code>99</code> is written to address <code>4</code>.</p>
<p>Parameters that an instruction writes to will <em>never be in immediate mode</em>.</p>
<p><em>Finally</em>, some notes:</p>
<ul>
<li>It is important to remember that the instruction pointer should increase by <em>the number of values in the instruction</em> after the instruction finishes. Because of the new instructions, this amount is no longer always <code>4</code>.</li>
<li>Integers can be negative: <code>1101,100,-1,4,0</code> is a valid program (find <code>100 + -1</code>, store the result in position <code>4</code>).</li>
</ul>
<p>The TEST diagnostic program will start by requesting from the user the ID of the system to test by running an <em>input</em> instruction - provide it <code>1</code>, the ID for the ship's air conditioner unit.</p>
<p>It will then perform a series of diagnostic tests confirming that various parts of the Intcode computer, like parameter modes, function correctly. For each test, it will run an <em>output</em> instruction indicating how far the result of the test was from the expected value, where <code>0</code> means the test was successful.  Non-zero outputs mean that a function is not working correctly; check the instructions that were run before the output instruction to see which one failed.</p>
<p>Finally, the program will output a <em>diagnostic code</em> and immediately halt. This final output isn't an error; an output followed immediately by a halt means the program finished.  If all outputs were zero except the diagnostic code, the diagnostic program ran successfully.</p>
<p>After providing <code>1</code> to the only input instruction and passing all the tests, <em>what diagnostic code does the program produce?</em></p>
</article>


In [10]:
class IntcodeComputer:
    def __init__(self, s: str) -> None:
        self.state = [int(i) for i in s.split(",")]
        self.reset_state = self.state[:]

    def run(self, inputs=None) -> list[int]:
        inputs = inputs if inputs else []
        output = []
        ip = 0
        while True:
            code = self.state[ip]
            opcode = code % 100
            code //= 100
            mode_a = code % 10
            code //= 10
            mode_b = code % 10
            code //= 10
            mode_c = code % 10

            if opcode == 1:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]

                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]

                ip += 1
                self.state[self.state[ip]] = a + b
                ip += 1
            elif opcode == 2:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]

                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]

                ip += 1
                self.state[self.state[ip]] = a * b
                ip += 1
            elif opcode == 3:
                ip += 1
                self.state[self.state[ip]] = inputs.pop()
                ip += 1
            elif opcode == 4:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                output.append(a)
                ip += 1
            elif opcode == 99:
                return output
            else:
                raise ValueError("1202 program alarm")

    def reset(self):
        self.state = self.reset_state[:]

    def __repr__(self) -> str:
        return ",".join(str(i) for i in self.state)


IntcodeComputer("1101,100,-1,4,0").run()

[]

In [11]:
from more_itertools import last


with open("../input/day5.txt") as f:
    puzzle = f.read()

print(f"Part I: {last(IntcodeComputer(puzzle).run([1]))}")

Part I: 16489636


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>16489636</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The air conditioner comes online! Its cold air feels good for a while, but then the TEST alarms start to go off. Since the air conditioner <span title="Honestly, I'm not sure what you expected.">can't vent its heat anywhere</span> but back into the spacecraft, it's actually making the air inside the ship <em>warmer</em>.</p>
<p>Instead, you'll need to use the TEST to extend the <a href="https://en.wikipedia.org/wiki/Spacecraft_thermal_control">thermal radiators</a>. Fortunately, the diagnostic program (your puzzle input) is already equipped for this.  Unfortunately, your Intcode computer is not.</p>
<p>Your computer is only missing a few opcodes:</p>
<ul>
<li>Opcode <code>5</code> is <em>jump-if-true</em>: if the first parameter is <em>non-zero</em>, it sets the instruction pointer to the value from the second parameter. Otherwise, it does nothing.</li>
<li>Opcode <code>6</code> is <em>jump-if-false</em>: if the first parameter <em>is zero</em>, it sets the instruction pointer to the value from the second parameter. Otherwise, it does nothing.</li>
<li>Opcode <code>7</code> is <em>less than</em>: if the first parameter is <em>less than</em> the second parameter, it stores <code>1</code> in the position given by the third parameter.  Otherwise, it stores <code>0</code>.</li>
<li>Opcode <code>8</code> is <em>equals</em>: if the first parameter is <em>equal to</em> the second parameter, it stores <code>1</code> in the position given by the third parameter.  Otherwise, it stores <code>0</code>.</li>
</ul>
<p>Like all instructions, these instructions need to support <em>parameter modes</em> as described above.</p>
<p>Normally, after an instruction is finished, the instruction pointer increases by the number of values in that instruction. <em>However</em>, if the instruction modifies the instruction pointer, that value is used and the instruction pointer is <em>not automatically increased</em>.</p>
<p>For example, here are several programs that take one input, compare it to the value <code>8</code>, and then produce one output:</p>
<ul>
<li><code>3,9,8,9,10,9,4,9,99,-1,8</code> - Using <em>position mode</em>, consider whether the input is <em>equal to</em> <code>8</code>; output <code>1</code> (if it is) or <code>0</code> (if it is not).</li>
<li><code>3,9,7,9,10,9,4,9,99,-1,8</code> - Using <em>position mode</em>, consider whether the input is <em>less than</em> <code>8</code>; output <code>1</code> (if it is) or <code>0</code> (if it is not).</li>
<li><code>3,3,1108,-1,8,3,4,3,99</code> - Using <em>immediate mode</em>, consider whether the input is <em>equal to</em> <code>8</code>; output <code>1</code> (if it is) or <code>0</code> (if it is not).</li>
<li><code>3,3,1107,-1,8,3,4,3,99</code> - Using <em>immediate mode</em>, consider whether the input is <em>less than </em><code>8</code>; output <code>1</code> (if it is) or <code>0</code> (if it is not).</li>
</ul>
<p>Here are some jump tests that take an input, then output <code>0</code> if the input was zero or <code>1</code> if the input was non-zero:</p>
<ul>
<li><code>3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9</code> (using <em>position mode</em>)</li>
<li><code>3,3,1105,-1,9,1101,0,0,12,4,12,99,1</code> (using <em>immediate mode</em>)</li>
</ul>
<p>Here's a larger example:</p>
<pre><code>3,21,1008,21,8,20,1005,20,22,107,8,21,20,1006,20,31,
1106,0,36,98,0,0,1002,21,125,20,4,20,1105,1,46,104,
999,1105,1,46,1101,1000,1,20,4,20,1105,1,46,98,99
</code></pre>
<p>The above example program uses an input instruction to ask for a single number.  The program will then output <code>999</code> if the input value is below <code>8</code>, output <code>1000</code> if the input value is equal to <code>8</code>, or output <code>1001</code> if the input value is greater than <code>8</code>.</p>
<p>This time, when the TEST diagnostic program runs its input instruction to get the ID of the system to test, <em>provide it <code>5</code></em>, the ID for the ship's thermal radiator controller. This diagnostic test suite only outputs one number, the <em>diagnostic code</em>.</p>
<p><em>What is the diagnostic code for system ID <code>5</code>?</em></p>
</article>

</main>


In [39]:
class IntcodeComputer:
    def __init__(self, s: str) -> None:
        self.state = [int(i) for i in s.split(",")]
        self.reset_state = self.state[:]

    def run(self, inputs=None) -> list[int]:
        inputs = inputs if inputs else []
        output = []
        ip = 0
        while True:
            code = self.state[ip]
            opcode = code % 100
            code //= 100
            mode_a = code % 10
            code //= 10
            mode_b = code % 10

            if opcode == 1:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]

                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]

                ip += 1
                self.state[self.state[ip]] = a + b
                ip += 1
            elif opcode == 2:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]

                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]

                ip += 1
                self.state[self.state[ip]] = a * b
                ip += 1
            elif opcode == 3:
                ip += 1
                self.state[self.state[ip]] = inputs.pop()
                ip += 1
            elif opcode == 4:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                output.append(a)
                ip += 1
            elif opcode == 5:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]
                if a != 0:
                    ip = b
                else:
                    ip += 1
            elif opcode == 6:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]
                if a == 0:
                    ip = b
                else:
                    ip += 1
            elif opcode == 7:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]
                ip += 1
                self.state[self.state[ip]] = 1 if a < b else 0
                ip += 1
            elif opcode == 8:
                ip += 1
                a = self.state[ip] if mode_a else self.state[self.state[ip]]
                ip += 1
                b = self.state[ip] if mode_b else self.state[self.state[ip]]
                ip += 1
                self.state[self.state[ip]] = 1 if a == b else 0
                ip += 1
            elif opcode == 99:
                return output
            else:
                raise ValueError("1202 program alarm")

    def reset(self):
        self.state = self.reset_state[:]

    def __repr__(self) -> str:
        return ",".join(str(i) for i in self.state)


tests = [
    {
        "name": "Example 1: Using position mode, consider whether the input is equal to 8",
        "program": "3,9,8,9,10,9,4,9,99,-1,8",
        "inputs": [8],
        "expected": 1,
    },
    {
        "name": "Example 2: Using position mode, consider whether the input is equal to 8",
        "program": "3,9,8,9,10,9,4,9,99,-1,8",
        "inputs": [3],
        "expected": 0,
    },
    {
        "name": "Example 3: Using position mode, consider whether the input is less than 8",
        "program": "3,9,7,9,10,9,4,9,99,-1,8",
        "inputs": [3],
        "expected": 1,
    },
    {
        "name": "Example 4: Using position mode, consider whether the input is less than 8",
        "program": "3,9,7,9,10,9,4,9,99,-1,8",
        "inputs": [8],
        "expected": 0,
    },
    {
        "name": "Example 5: Using immediate mode, consider whether the input is equal to 8",
        "program": "3,3,1108,-1,8,3,4,3,99",
        "inputs": [8],
        "expected": 1,
    },
    {
        "name": "Example 6: Using immediate mode, consider whether the input is equal to 8",
        "program": "3,3,1108,-1,8,3,4,3,99",
        "inputs": [9],
        "expected": 0,
    },
    {
        "name": "Example 7: Using immediate mode, consider whether the input is less than 8",
        "program": "3,3,1107,-1,8,3,4,3,99",
        "inputs": [9],
        "expected": 0,
    },
    {
        "name": "Example 8: Using immediate mode, consider whether the input is less than 8",
        "program": "3,3,1107,-1,8,3,4,3,99",
        "inputs": [1],
        "expected": 1,
    },
    {
        "name": "Example 9: Jump tests that take an input, then output 0 if the input was zero or 1 using position mode",
        "program": "3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9",
        "inputs": [0],
        "expected": 0,
    },
    {
        "name": "Example 10: Jump tests that take an input, then output 0 if the input was zero or 1 using position mode",
        "program": "3,12,6,12,15,1,13,14,13,4,13,99,-1,0,1,9",
        "inputs": [1],
        "expected": 1,
    },
    {
        "name": "Example 11: Jump tests that take an input, then output 0 if the input was zero or 1 using immediate mode",
        "program": "3,3,1105,-1,9,1101,0,0,12,4,12,99,1",
        "inputs": [0],
        "expected": 0,
    },
    {
        "name": "Example 12: Jump tests that take an input, then output 0 if the input was zero or 1 using immediate mode",
        "program": "3,3,1105,-1,9,1101,0,0,12,4,12,99,1",
        "inputs": [1],
        "expected": 1,
    },
    {
        "name": "Example 13: larger example",
        "program": "3,21,1008,21,8,20,1005,20,22,107,8,21,20,1006,20,31,1106,0,36,98,0,0,1002,21,125,20,4,20,1105,1,46,104,999,1105,1,46,1101,1000,1,20,4,20,1105,1,46,98,99",
        "inputs": [7],
        "expected": 999,
    },
    {
        "name": "Example 14: larger example",
        "program": "3,21,1008,21,8,20,1005,20,22,107,8,21,20,1006,20,31,1106,0,36,98,0,0,1002,21,125,20,4,20,1105,1,46,104,999,1105,1,46,1101,1000,1,20,4,20,1105,1,46,98,99",
        "inputs": [8],
        "expected": 1000,
    },
    {
        "name": "Example 11: larger example",
        "program": "3,21,1008,21,8,20,1005,20,22,107,8,21,20,1006,20,31,1106,0,36,98,0,0,1002,21,125,20,4,20,1105,1,46,104,999,1105,1,46,1101,1000,1,20,4,20,1105,1,46,98,99",
        "inputs": [9],
        "expected": 1001,
    },
]


def test_runner(program, inputs):
    return last(IntcodeComputer(program).run(inputs))


run_tests_params(test_runner, tests)


Test Example 1: Using position mode, consider whether the input is equal to 8 passed, for test_runner.
Test Example 2: Using position mode, consider whether the input is equal to 8 passed, for test_runner.
Test Example 3: Using position mode, consider whether the input is less than 8 passed, for test_runner.
Test Example 4: Using position mode, consider whether the input is less than 8 passed, for test_runner.
Test Example 5: Using immediate mode, consider whether the input is equal to 8 passed, for test_runner.
Test Example 6: Using immediate mode, consider whether the input is equal to 8 passed, for test_runner.
Test Example 7: Using immediate mode, consider whether the input is less than 8 passed, for test_runner.
Test Example 8: Using immediate mode, consider whether the input is less than 8 passed, for test_runner.
Test Example 9: Jump tests that take an input, then output 0 if the input was zero or 1 using position mode passed, for test_runner.
Test Example 10: Jump tests that t

In [40]:
print(f"Part II: {test_runner(puzzle, [5])}")

Part II: 9386583


<link href="style.css" rel="stylesheet"></link>
<main class="read-aloud">

<p>Your puzzle answer was <code>9386583</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>

</main>
